### PR TITLE
[Merged by Bors] - feat: convenient meta declarations

### PR DESCRIPTION
**Pull request by Komyyy**
These declarations are picked from #5938.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by digama0** [comment]
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

File: Mathlib/Lean/Expr/Basic.lean
Comment (Vierkantor) [inline Mathlib/Lean/Expr/Basic.lean:476]:
```suggestion
            (m!"ill-formed expression, {sc} is the {pinfo.i}-th projection function" ++
              m!" but {sarg} does not have enough arguments")
```

File: Mathlib/Lean/Meta/Basic.lean
Comment (Vierkantor) [inline Mathlib/Lean/Meta/Basic.lean:7]:
```suggestion
import Lean.Meta.AppBuilder
import Lean.Meta.Basic
```

File: Mathlib/Lean/Expr/Basic.lean
Comment (Vierkantor) [inline Mathlib/Lean/Expr/Basic.lean:197]:
```suggestion
/-- Invariant: `i : ℕ` should be less than the size of `as : Array Expr`. -/
private def getAppAppsAux : Expr → Array Expr → Nat → Array Expr
```

**Review by Vierkantor — APPROVED** [review]
I have some minor remarks, LGTM otherwise. But it feels like this is slightly above my level of expertise.

**Comment by Vierkantor** [comment]
You know what, I'm going to assume this needs no substantial revision:

maintainer merge

File: Mathlib/Lean/Expr/Basic.lean
Comment (semorrison) [inline Mathlib/Lean/Expr/Basic.lean:203]: What is this for? It seems an unlikely thing to want in general?

File: Mathlib/Lean/Expr/Basic.lean
Comment (semorrison) [reply Mathlib/Lean/Expr/Basic.lean:531]: Do you actually want this, or do you want a `MetaM` level `push_neg`?

File: Mathlib/Lean/Expr/Basic.lean
Comment (Komyyy) [reply Mathlib/Lean/Expr/Basic.lean:203]: This is used in the `cc` tactic.

File: Mathlib/Lean/Expr/Basic.lean
Comment (Komyyy) [reply Mathlib/Lean/Expr/Basic.lean:531]: I actually want this for the `cc` tactic.

File: Mathlib/Lean/Expr/Basic.lean
Comment (kmill) [inline Mathlib/Lean/Expr/Basic.lean:479]: As I was trying to evaluate this, I added some comments, expanded `sfds` to `sfields` (it's better not to use cryptic names), and reduced indentation (which not only makes it a bit easier to follow I think, but it also is more efficient).

```suggestion
/-- If `e` is a projection of the structure constructor, reduce the projection.
Otherwise returns `none`. If this function detects that expression is ill-typed, throws an error.

For example, given `Prod.fst (x, y)`, returns `some x`. -/
def reduceProjStruct? (e : Expr) : MetaM (Option Expr) := do
  let .const cname _ := e.getAppFn | return none
  let some pinfo ← getProjectionFnInfo? cname | return none
  let args := e.getAppArgs
  if ha : args.size = pinfo.numParams + 1 then
    -- The last argument of a projection is the structure.
    let sarg := args[pinfo.numParams]'(ha ▸ pinfo.numParams.lt_succ_self)
    -- Check that the structure is a constructor expression.
    unless sarg.getAppFn.isConstOf pinfo.ctorName do
      return none
    let sfields := sarg.getAppArgs
    -- The ith projection extracts the ith field of the constructor
    let sidx := pinfo.numParams + pinfo.i
    if hs : sidx < sfields.size then
      return some (sfields[sidx]'hs)
    else
      throwError m!"ill-formed expression, {cname} is the {pinfo.i + 1}-th projection function {
          ""}but {sarg} does not have enough arguments"
  else
    return none
```

File: Mathlib/Lean/Meta/Basic.lean
Comment (kmill) [inline Mathlib/Lean/Meta/Basic.lean:58]: I'm not so sure about this, since it's just doing what the `isDefEq` documentation suggests doing, and it's just saving a few characters while adding another function to the API.

If you do want it, we need to make the documentation very clear about what it does.

```suggestion
/-- `pureIsDefEq e₁ e₂` is short for `withNewMCtxDepth <| isDefEq e₁ e₂`.
Determines whether two expressions are definitionally equal to each other
when metavariables are not assignable. -/
```

File: Mathlib/Lean/Meta/Basic.lean
Comment (kmill) [inline Mathlib/Lean/Meta/Basic.lean:63]: We need to know from the documentation how it's similar, so here's a suggestion.

```suggestion
/-- `mkRel n lhs rhs` is `mkAppM n #[lhs, rhs]`, but with optimizations for `Eq` and `Iff`. -/
```

File: Mathlib/Tactic/Relation/Rfl.lean
Comment (kmill) [inline Mathlib/Tactic/Relation/Rfl.lean:39]: If you're intending to use `isReflRel` with `mkRel`, then there's a bug because there's no reason to assume that this constant takes only implicit arguments followed by two explicit arguments.

We could fix `mkRel` so that it uses `mkAppOptM` and feeds in `lhs` and `rhs` as the last two explicit arguments. Or, CC could document this limitation with its ability to handle reflexive/symmetric relations and we can fix it later.

Name-wise, I think it shouldn't use `is` since it doesn't return a `Bool`, and it should end in `?` since it returns an `Option`. Perhaps `Lean.Expr.relSidesIfRefl?`, though I'm not really sure.

**Review by kmill — COMMENTED** [review]
It would be good to have some tests that these functions actually work. After you address these comments, and if you don't write any yourself, I can add some tests before merging.

**Comment by Komyyy** [comment]
@kmill I would like write the tests. Thank you.

**Comment by Komyyy** [comment]
@kmill 
✖ I would like write the tests. Thank you.
◯ I would like you to write the tests. Thank you.
Sorry.

**Comment by kmill** [comment]
Thanks!

bors r+

**Pull request closed without merging**
